src/OptimalFilter.jl

```diff
@@ -0,0 +1,396 @@
+
+    normalization_factor = 1.0 / sqrt(model_params.nx_ext * model_params.ny_ext)
+    transformed_vector .= fft(reshape(vector, model_params.nx_ext, model_params.ny_ext))[:] .* normalization_factor
+
+end
+
+# Normalized inverse two-dimension discrete Fourier transofrm normalized by sqrt(n1_bar).
+# Operates on the 2d data stored as a matrix.
+# From Dietrich & Newsam 96 in text following equation 12
+function normalized_inverse_2d_fft!(transformed_array::AbstractMatrix{T}, array::AbstractMatrix{S}, model_params) where {T,S}
+
+    normalization_factor = sqrt(model_params.nx_ext * model_params.ny_ext)
+    transformed_array .= ifft(array) .* normalization_factor
+
+end
+
+# Normalized inverse two-dimension discrete Fourier transofrm normalized by sqrt(n1_bar).
+# Operates on the 2d data stored as a vector.
+# From Dietrich & Newsam 96 in text following equation 12
+function normalized_inverse_2d_fft!(transformed_vector::AbstractVector{T}, vector::AbstractVector{S}, model_params) where {T,S}
+
+    normalization_factor = sqrt(model_params.nx_ext * model_params.ny_ext)
+    transformed_vector .= ifft(reshape(vector, model_params.nx_ext, model_params.ny_ext))[:] .* normalization_factor
+
+end
+
+# Decomposition of R11, equation 12 of Deitrich and Newsam
+function WΛWH_decomposition!(transformed_array::AbstractMatrix{T}, array::AbstractMatrix{T},
+                             offline_matrices::OfflineMatrices, params) where T
+
+    @assert size(array) == (params.nx, params.ny)
+
+    extended_array = zeros(ComplexF64, params.nx_ext, params.ny_ext)
+
+    extended_array[1:params.nx, 1:params.ny] .= array
+
+    normalized_2d_fft!(extended_array, extended_array, params)
+
+    # Here we do an element-wise multiplication of the extended_array with the vector Lambda. This is identical to
+    # Diagonal(Lambda) * extended_array[:], but avoids flattening and reshaping extended_array.
+    normalized_inverse_2d_fft!(extended_array, reshape(offline_matrices.Lambda, params.nx_ext, params.ny_ext).*extended_array, params)
+
+    transformed_array .= real.(@view(extended_array[1:params.nx, 1:params.ny]))
+
+end
+
+# Get a vector of values of field at positions of [stations.ist, stations.jst]
+function get_values_at_stations(field::AbstractMatrix{T}, stations) where T
+
+    values = zeros(T, length(stations.ist))
+
+    for (num,(i,j)) in enumerate(zip(stations.ist, stations.jst))
+        values[num] = field[i,j]
+    end
+
+    return values
+
+end
+
+# Allocate and compute matrices that do not depend on time-dependent variables (height and observations).
+function init_offline_matrices(model_params, filter_params, stations)
+
+    n1 = model_params.nx * model_params.ny # number of elements in original grid
+    n1_bar = model_params.nx_ext * model_params.ny_ext     # number of elements in extended grid
+
+    F = Float64
+    C = ComplexF64
```

Comment:
If you want to infer `C` from `F`, you can do
```suggestion
    C = complex(F)
```
for when we can use different float types